### ISSUE TITLE
Create a widgets.Button class

### ISSUE THREAD

**Issue by myk002** [DFHack/dfhack#3382]
that renders reasonably in both ascii and graphics mode

should be able to take a list of tiles (or a tile start position and a layout specifier for w,h?) for buttons that have icons

should be able to replace the janky button on the `buildingplan` item selection dialog, which is just text and a background (but we can change the style there if we want a better border

this is part of #3257

**Comment by myk002** [comment]
@TaxiService could you possibly attach mocks here?
@johncosker do you want to take ownership for implementing?

**Comment by johncosker** [comment]
Yeah, I've already got a little bit of prototyping done that creates a clickable button.

I was going to create a `ButtonGroup` class as well with options to make groups of buttons easier to implement and use.

I think for button graphics we could come up with a 'standard' layout for button graphics png files, maybe 3 rows for 'unpressed', 'pressed', and 'hover' (if we want to, could be optional), and then N columns for each button in the graphics file.

The button graphics Taxi provided me are '4x3' but we probably should support any dimension, as long as it's divisible into the 8x12 pixel text character dims.

**Comment by TaxiService** [comment]
here's a first batch of things to consider.

- some Buttons will be "sprite-based", like the ones in `gui/design`.
![image](https://github.com/DFHack/dfhack/assets/5853667/68dcb029-d112-4dc0-a398-99bfbe78ad04)
Each of these will have a custom sprite for its states. 
These Buttons' dimensions depend on its associated sprites.

- ...other buttons will be "text-based", like the `?: Quickstart Guide` button on title screen.
![image](https://github.com/DFHack/dfhack/assets/5853667/28a64800-7fde-4be3-a200-32e19be71a64)
For these, we'll define some appropriately textured FRAMEs, and switch between them depending on states.
These Buttons' dimensions depend on whatever text we choose to put in them.

here is an example of how one of these spritesheets could look like. 
![button_sheet_1](https://github.com/DFHack/dfhack/assets/5853667/37e1e28e-09d1-4994-90bc-3eee4cb5870c)
and here's an explanation of why there are so many sprites...

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

thinking about it... a button's various states could maybe be abstracted/described like this:

States:
- `off` (default state)
This is a button that is either connected to an option which is not active, or 
a button that does _not_ turn on at all. (like the `?: Quickstart Guide`)
- `on`
This is a button that is connected to an option that is active.
- `disabled`
This is a button that can not be clicked.
Implementation is optional, but i think something like this could turn out useful.

The `on` and `off` states can then have the following sub-states:
- `unhovered` (default sub-state)
The mouse is nowhere near this button.
- `hovered`
This is a button on which the mouse cursor has been moved over.
- `pressed` 
This is a button that is being clicked right now. It's a transition state between off/on, and stays active as long as mouse1 is held down.
Implementation is optional, but having this would make the buttons feel *much better*, and other advantages. (†)

(`disabled` does not need hovered/pressed since it can't be interacted with)
(also, hovered/pressed are only going to be visible for a mouse user. A button toggled with a hotkey would instantly switch from off-unhovered to on-unhovered, or vice versa)

here is a table of the various states, and a couple of gifs to help visualize how it all ties together.
![buttons_table](https://github.com/DFHack/dfhack/assets/5853667/b3de7210-972a-4474-8661-efea59e808e5)

![buttonspremium1](https://github.com/DFHack/dfhack/assets/5853667/029a859e-3534-47a9-9a63-0eab52870f3d) ![buttonsclassic1](https://github.com/DFHack/dfhack/assets/5853667/d5abe2c0-35f6-40db-91f3-506857fb89c0)

I'm sure i'm forgetting something... but whatever. This is some stuff to think about, for now. °-°
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
(†) : by "other advantages" i mean that we could implement a behaviour like ![this](https://github.com/DFHack/dfhack/assets/5853667/70c278ec-7e6e-4fee-99c6-1d82c0f77437).
In which, as long as left-click is held down, the Button does not change state.
State is only changed if left-click is released _while the cursor is still on the button_. 
So if you have second thoughts, you can move the mouse away and release left-click without committing the change!

(one disadvantage is that it's 1 additional sub-state per state. 
...and, to be honest, we would actually need _2 sub-states_. 🙄 `hovered-pressed` and `unhovered-pressed`...
I havent included this fourth state in the mockups, but if we want our UX to be of the highest quality, we should do this)

**Comment by TaxiService** [comment]
moreover, look at the states' table i posted above. Notice how, for example, the `Off` state is _grey_, but becomes _yellow_ while `pressed`.

To allow for more freedom for choosing which colors the Button uses in its different states, we should not "hard-code" the _yellow_ pressed state within the _grey_ texture. (what if the active state is green instead?)

So, I'm thinking that our Button class should be expecting 2 "frame values" or something, one for its `Off` state, and one for its `On` state.

Basically, these "frame values" point to a spritesheet with all the possible permutations of a button of a single color.
![image](https://github.com/DFHack/dfhack/assets/5853667/c1000ad2-a4c7-478a-abc5-061165e55154)

But note that the "pressed" sub-states are going to be used by a Button that, when clicked, is _turning into this color!_
Like in the table from previous post; if our `off-frame` was _gray_ and `on-frame` was _yellow_, while a _gray_ OFF button is being pressed, it would show the _yellow_ sheet's sprite; and vice versa.

All of this crap would let us easily add more colors, and also give us lots of freedom in how our buttons are styled.
![image](https://github.com/DFHack/dfhack/assets/5853667/e4a8839d-f431-4f0e-becb-8f35c82393d9)

This also makes it easier for the end-user to mod these graphics later.
Like, if they want to change the red buttons, they can change all of them from the same texture, etc.


PS: also, in these 'sheets i've included both `pressed-hovered` and `pressed-unhovered` states 🙄 i still dont know if any of the _hovered_ things are going to make it in, but here they are.

**Comment by TaxiService** [comment]
[button-frames.zip](https://github.com/DFHack/dfhack/files/11529848/button-frames.zip) - here's a zipped folder with these files in it.

![image](https://github.com/DFHack/dfhack/assets/5853667/e8846692-7289-41c5-a05c-5f958f7df37b)

I've included two 'sheets for `disabled`, one with just the unpressed-unhovered state, and one with all states (even if we should probably never see them... but at least it's another 12x6 sheet and the code for the make-button-frames function could be reused as-is?)

**Comment by TaxiService** [comment]
also it is worth nothing that if we were able to _overlay transparent textures on top of other textures_, we could halve the size of these spritesheets by using something like this for the `hovered` states.

![buttons_hovered](https://github.com/DFHack/dfhack/assets/5853667/dfe0839a-ee16-4935-a0ce-6cd976706caa)


also worthy of note, if we were able to _rotate sprites by 180 degrees_ we could halve the spritesheets again, because the `pressed` states are just the normal button upside-down!

idk which of these is harder, and if any are possible, but i had to mention this.

**Comment by johncosker** [comment]
wow there's a lot here to catch up on, thanks for the button frames, i'll try to get more done on this

**Comment by johncosker** [comment]
@TaxiService hey could you create a spritesheet for these types of buttons with all the states?

![image](https://github.com/DFHack/dfhack/assets/4482627/e0b643de-4ef6-4af4-8a4f-45c79e8dc0f7)

I'm working on the sprite based buttons first, then I'll look at the text button

**Comment by TaxiService** [comment]
@johncosker 
[button-design.zip](https://github.com/DFHack/dfhack/files/11560014/button-design.zip)

here, there's both a single spritesheet and individual spritesheets per button
![image](https://github.com/DFHack/dfhack/assets/5853667/ec18bd49-61ea-476f-bb61-d3a711236375)

**Comment by johncosker** [comment]
Awesome thanks, yeah probably won't need disable yet, although it would probably help to figure out where it'd go in the layout and standardize it, would it just be the fifth column in the button_design.png?

**Comment by TaxiService** [comment]
damn... i don't know. :thinking: I mean, i was thinking the same last night, but today i'm like... "disabled is a third _state_, so it should have its own _row_"

nevertheless, here's a 'sheet with the disabled state tacked on at the end of the first row per button. 🤔
we can work with this, but if you have any other idea lemme know. We could theoretically even have a dedicated 'sheet only for disabled sprites or something; that would let us not have the magenta empty space.

![button_design2](https://github.com/DFHack/dfhack/assets/5853667/c62c6e5c-be5a-4de9-9d34-d791527d2a98)

**Comment by ab9rf** [comment]
We do have a lot of buttons, but we don't actually have a "generic button", but I still think this can be closed
